snowpack/src/build/optimize.ts

```diff
@@ -0,0 +1,538 @@
+    .each((_, elem) => {
+      const scriptRoot = root(elem);
+      const scriptFile = path.resolve(file, '..', scriptRoot.attr('src')!);
+      const scriptContent = readFileSync(scriptFile, 'utf-8');
+      scriptRoot.text(scriptContent);
+      scriptRoot.removeAttr('src');
+      scriptRoot.removeAttr('snowpack-inline');
+      unlinkSync(scriptFile);
+    });
+  getLinks('stylesheet')
+    .filter('[snowpack-inline]')
+    .each((_, elem) => {
+      const linkRoot = root(elem);
+      const styleFile = path.resolve(file, '..', linkRoot.attr('href')!);
+      const styleContent = readFileSync(styleFile, 'utf-8');
+      const newStyleEl = root('<style></style>');
+      newStyleEl.text(styleContent);
+      linkRoot.after(newStyleEl);
+      linkRoot.remove();
+      unlinkSync(styleFile);
+    });
+}
+
+/** Add new bundled CSS files to the HTML entrypoint file, if not already there. */
+function addNewBundledCss(
+  htmlData: ScannedHtmlEntrypoint,
+  manifest: ESBuildMetaManifest,
+  baseUrl: string,
+): void {
+  if (!manifest.outputs) {
+    return;
+  }
+  for (const key of Object.keys(manifest.outputs)) {
+    if (!key.endsWith('.css')) {
+      continue;
+    }
+    const scriptKey = key.replace('.css', '.js');
+    if (!manifest.outputs[scriptKey]) {
+      continue;
+    }
+    const hasCssImportAlready = htmlData
+      .getLinks('stylesheet')
+      .toArray()
+      .some((v) => v.attribs.href.includes(removeLeadingSlash(key)));
+    const hasScriptImportAlready = htmlData
+      .getScripts()
+      .toArray()
+      .some((v) => v.attribs.src.includes(removeLeadingSlash(scriptKey)));
+
+    if (hasCssImportAlready || !hasScriptImportAlready) {
+      continue;
+    }
+    const linkHref = removeTrailingSlash(baseUrl) + addLeadingSlash(key);
+    htmlData.root('head').append(`<link rel="stylesheet" href="${linkHref}" />`);
+  }
+}
+
+/**
+ * Traverse the entrypoint for JS scripts, and add preload links to the HTML entrypoint.
+ */
+function preloadEntrypoint(
+  htmlData: ScannedHtmlEntrypoint,
+  manifest: ESBuildMetaManifest,
+  config: SnowpackConfig,
+): void {
+  const {root, getScripts} = htmlData;
+  const preloadScripts = getScripts()
+    .map((_, elem) => elem.attribs.src)
+    .get()
+    .filter(isTruthy);
+  const collectedDeepImports = new Set<string>();
+  for (const preloadScript of preloadScripts) {
+    const normalizedPreloadScript = removeLeadingSlash(
+      path.posix.resolve('/', removeLeadingSlash(preloadScript)),
+    );
+    collectDeepImports(normalizedPreloadScript, manifest, collectedDeepImports);
+  }
+  const baseUrl = config.buildOptions.baseUrl;
+  for (const imp of collectedDeepImports) {
+    const preloadUrl = (baseUrl ? removeTrailingSlash(baseUrl) : '') + addLeadingSlash(imp);
+    root('head').append(`<link rel="modulepreload" href="${preloadUrl}" />`);
+  }
+}
+
+/**
+ * Given any set of user-input as entrypoints,
+ */
+async function resolveEntrypoints(
```

Comment:
More of a DX observation than anything: for my project, I have `src/index.js`. So in my config, I specified `entrypoints: ['./src/index.js']`. But it threw a “not found“ error. I guess I have to specify `./_dist_/index.js` instead here?

Reply:
Ah I see—it wants HTML entrypoints. Hm, similar thing, though—if I specify `['./public/index.html']` or  `['./index.html']` both throw a “not found“ error, so I’m not sure how to use this

Reply:
It should work with both JS & HTML entrypoints, but it looks like there's some bug here. since the 'auto' scanner returns absolute paths, i guess it needs absolute file paths to the final `build/` output files? Either way, will fix to support relative files 👍

Reply:
> It should work with both JS & HTML entrypoints, but it looks like there's some bug here. since the 'auto' scanner returns absolute paths, i guess it needs absolute file paths to the final `build/` output files? Either way, will fix to support relative files 👍

Gotcha. Yeah if I specify `.js` I get a “only HTML is supported” message right now

Reply:
Also: need to fix the "seeing modulepreloads when only trying to bundle"  issue